### ISSUE TITLE
Reconfigurable hardware platform support

### ISSUE THREAD

**Issue by danxn** [runtimejs/runtime#45]
Sorry, if this is an offtopic, i'll move it to maillist if it's so. This just common suggestions about "full metal runtime.js/nodejs" related to this #42  and this #43  issues.

Besides all we all like JS because it is dynamic language. We can modify runtime envieronment during programm execution, create functionality on the fly, modify it, pass functions as parameters to the functions and so on. And now JS-driven technologies comes close to hardware layer of computing systems. Why not to try to combine dynamic software with dynamic hardware? I mean such technologies, like [FPGA (Field-programmable gate array)](http://en.wikipedia.org/wiki/Field-programmable_gate_array) and [PSoC (Programmable System on a Chip)](http://en.wikipedia.org/wiki/PSoC) i.e. [Reconfigurable computing](http://en.wikipedia.org/wiki/Reconfigurable_computing).

First I would like to quotate article about Intel experience in using combined server systems with Intel Xeon and FPGA modules: [Intel unveils new Xeon chip with integrated FPGA, touts 20x performance boost](http://www.extremetech.com/extreme/184828-intel-unveils-new-xeon-chip-with-integrated-fpga-touts-20x-performance-boost).

> Late yesterday, Intel quietly announced one of the biggest ever changes to its chip lineup: It will soon offer a new type of Xeon CPU with an integrated FPGA. This new Xeon+FPGA chip will fit in the standard E5 LGA2011 socket, but the integrated FPGA will allow each chip to be customized to specific workloads.

> What’s the purpose of this new Xeon+FPGA product? In the words of Intel: “The FPGA provides our customers a programmable, high performance coherent acceleration capability to turbo-charge their critical algorithms.” Intel estimates that the Xeon+FPGA will see massive performance boosts in the 20x range (for code executed on the FPGA instead of a conventional x86 CPU — but obviously there will be big overall speedups as bottlenecks are removed. The other advantage is that workloads change — so if your critical algorithms change, or your whole company pivots, the FPGA can be repurposed without having to buy lots of new hardware.

__So, for server applications this approach shows very good experimental results. With help of reconfigurable and programmable hardware we can implement device drivers, media codecs, network protocols, OS modules, crytical applications algorythms, additional devices and many other things.__

Probably it is possible to make even "full-metal" V8 engine as CPU core, but probably this has no much sense, because hardware processor and dynamic interpreter is quite different things. I've herd about hardware implementations of systems like Forth, for example (it is also programming language, interpreter and operating system in one), when forth interpreter implemented as CPU core, but, looks like this is not very popular application. So it is possible to use all power of programmable [MPSoC]s (http://en.wikipedia.org/wiki/MPSoC) in combination with compiled V8 core and additional core and applications modules, implemented programmatically or as a hard logic (as additional PCI devices, for example, or as modules with direct access to builtin CPU cores cache).

Intel used in their combined systems FPGA from Altera. As I can see more suitable for runtime.js system is [All Programmable MPSoCs](http://www.xilinx.com/products/silicon-devices/soc.html) from Xilinx.

![image](https://cloud.githubusercontent.com/assets/12805166/8116880/65c1fb8a-108d-11e5-826f-29af495bff97.png)

There are two series of Programmable SoCs form Xilinx: currently available [Zynq-7000 AP SoC](http://www.xilinx.com/products/silicon-devices/soc/zynq-7000.html) and newest series - [Zynq® UltraScale+™ MPSoC](http://www.xilinx.com/products/silicon-devices/soc/zynq-ultrascale-mpsoc.html). The short description with architectures schemas of UltraScale can be found in this article - [Xilinx Introduces Zynq UltraScale+ MPSoC with Cortex A53 & R5 Cores, Ultrascale FPGA] (http://www.cnx-software.com/2015/03/05/xilinx-introduces-zynq-ultrascale-mpsoc-with-cortex-a53-r5-cores-ultrascale-fpga/), technical details can be found here [Zynq UltraScale+ MPSoC
Product Tables and Product Selection Guide](http://www.xilinx.com/publications/prod_mktg/zynq-ultrascale-plus-product-table-product-selection-guide.pdf)

Here I'll describe in details some Zynq-7000 applications.

### Common schema of Xilinx [ultrascale architecture](http://www.xilinx.com/products/technology/ultrascale-mpsoc.html):

![image](https://cloud.githubusercontent.com/assets/12805166/8116995/d21915d4-108d-11e5-9fe0-3f2f1ae91485.png)

### Zynq-7000 All Programmable SoC Overview

> A high-level block diagram of the Zynq-7000 AP SoC is shown in Figure:

![zynq-7000](https://cloud.githubusercontent.com/assets/12805166/8117013/f3fe4638-108d-11e5-8bd8-f2a328962b1f.png)

> Each ARM Cortex-A9 processor has two 32 KB built-in level-1 caches for instructions and data, respectively. Another 512 KB on-chip level-2 cache is shared by the two processors. The snoop control unit (SCU) maintains the coherency of the caches in the two processors. The interconnection between the processor system and programmable logic is achieved through nine distinct AXI ports. The S_AXI_GP slave ports are typically used by the programmable logic that needs to access the processor system peripherals, while the M_AXI_GP master ports are mainly used by the processor system to access the register maps built in the programmable logic. The S_AXI_HP slave ports provide an efficient way for the programmable logic to access an external DDR memory or the 256 KB on-chip memory, while, for latency sensitive applications, the accelerator coherent port S_AXI_ACP offers direct accesses to the caches via the SCU. For a more detailed description of the Zynq-7000 AP SoC.

[Wireless Base Station ZUC Block Cipher Implementation on Zynq-7000 AP SoC] (http://www.xilinx.com/support/documentation/white_papers/wp437-ZUC-Cipher-Zynq.pdf)

This AP SoCs now available as SoM (System on Module) from Avnet - [Zynq-7000 All Programmable SoC Systems](http://www.em.avnet.com/en-us/design/drc/Pages/Zynq-7000-All-Programmable-SoC-Systems.aspx):

![image](https://cloud.githubusercontent.com/assets/12805166/8117052/3017ecc8-108e-11e5-9b22-eaf6801c117d.png)

From my point of view the most suitable for small cloud-enabled server module for runtime.js is MicroZed:

_MicroZed™ is a low-cost System-On-Module, or SOM that is based on the Xilinx Zynq®-7000 All Programmable (AP) SoC. In addition to the Zynq-7000 AP SoC, the module contains the common functions and interfaces required to support the core of most SoC designs, including memory, configuration, Ethernet, USB, and clocks. On the bottom side of the module, MicroZed contains two 100-pin I/O headers that provide connection to two I/O banks on the programmable logic (PL) side of the Zynq-7000 AP SoC device. When plugged onto a user designed baseboard or carrier card, these 100-pin connectors provide connectivity between the Zynq-7000 AP SoC PL I/Os and the user circuits on the carrier card. MicroZed also includes onboard power regulation that can support a single 5 V to 12 V input._

![image](https://cloud.githubusercontent.com/assets/12805166/8117083/636df54a-108e-11e5-8ce3-f2db407d8aaf.png)

##### Key Features

- __SoC__
  - XC7Z010-1CLG400 or XC7Z020-1CLG400
- __Memory__
  - 1 GB of DDR3 SDRAM
  - 128 Mb of QSPI Flash
  - Micro SD card interface
- __Communications__
  - 10/100/1000 Ethernet
  - USB 2.0
  - USB-UART
- __User I/O (via dual board-to-board connectors)__
  - 7Z010 Version: 100 User I/O (50 per connector) Configurable as up to 48 LVDS pairs or 100 single-ended I/O
  - 7Z020 Version: 115 User I/O (58/57 per connector) Configurable as up to 55 LVDS pairs or 115 single-ended I/O
- __Other__
  - 2x6 Digilent Pmod® compatible interface providing 8 PS MIO connections for user I/O
  - Xilinx PC4 JTAG configuration port
  - PS JTAG pins accessible via Pmod or I/O headers
  - 33.33 MHz oscillator
  - User LED and push switch
- __Software__
  - Linux BSP and reference design
- __Mechanical__
  - 4 inches x 2.25 inches (102 mm x 57 mm)

I think this could be very intresting perspective direction of porting runtime.js and nodejs/nodeos, because due to reconfigurable (dynamic) properties of this hardware it is possible to achieve very interesting results in server system performance, similar to Intel's ones.

Zynq-7000 contains Dual-Core 32-bit ARM Cortex-A9, Zynq UltraSCALE carries Quad-Core 64-bit ARM Cortex -A53.

**Comment by piranna** [comment]
Interesting, although I don't believe it correspond to runtime.js or NodeOS... at least directly. Someone could write a npm module to rewrite fpga and wrap it with some functions so it gets "hardware acelerated Javascript" similar to what's currently done with JIT (only a lower level).

**Comment by iefserge** [comment]
Interesting, didn't know that this exists! Dynamic hardware sounds like something V8 could use internally to speed up JavaScript. Probably not much runtime.js can do here.

**Comment by danxn** [comment]
@piranna
> although I don't believe it correspond to runtime.js or NodeOS... at least directly

Well, this is corresponded to processors architectures supported by node.js (ARM) - in other words the same issue, as about porting to Raspberry PI :) Because current available AP SoCs carrying ARM CPU cores. If runtime.js or nodeos/nodejs will support this architecture it will be first step to full support of reconfigurable hardware support.

@piranna
> Someone could write a npm module to rewrite fpga and wrap it with some functions so it gets "hardware acelerated Javascript" similar to what's currently done with JIT (only a lower level).

Yes, it's not very easy and it would be great if necessary toolset will be written in runtime.js/nodejs and can be used for reconfiguration of hardware, but for now appropriate tools alredy exists and works well - 

### System Hardware Definition
> The development flow for the Zynq-7000 AP SoC consists of two major steps. The
system hardware definition step defines the various components of the system and
how they are connected together. The Xilinx Platform Studio (XPS) [Ref 10] is used to
configure the main hardware parameters of each component and IP:
> - Clock source and frequency
> - AXI interconnect ports and base addresses
> - Connection to interrupt controllers
> - Connection to external input/output pins

> For Xilinx IP cores, including those generated by Vivado HLS, the interconnections
and bus configurations are fully automated and customizable. After the processor
system has been defined with XPS, Xilinx PlanAhead™ [Ref 11] can automatically
generate a top-level RTL wrapper that can be synthesized by the ISE® or Vivado tools.
The resulting BIT file is finally downloaded to the Zynq-7000 AP SoC, which then
behaves as a fully customized application-specific device under software control.

**Comment by danxn** [comment]
@iefserge 

> Probably not much runtime.js can do here.

Ah, don't worry :) All what is needed from runtime.js is ARM support :)

**Comment by danxn** [comment]
@iefserge 
> Interesting, didn't know that this existed! Dynamic hardware sounds like something V8 could use internally to speed up JavaScript.

Yes, exactly, but V8 compiles this code and it remains software. With help of FPGA parts embedded in reconfigurable hardware like Zynq-7000 AP SoC it is possible for example (theoretically) to convert one of runtime.js modules, written in JS, to hardware module which will be connected to runtime.js core via PCI bus, or via mapped memory registers (like additional device or part of CPU - with appropriate performance). As well as additional simple hardware and drivers ("soft hardware") for devices which utilizes only GPIO, for example. And this can be changed after reconfiguring/reloading to another configuration of hard/soft parts of system.

**Comment by piranna** [comment]
I still think this could be done as a npm module and don't have a direct relationship with runtime.js beyond the fact that it could provide some modules to access the direct hardware (that could also be npm modules available for other Javascript environments...).

**Comment by danxn** [comment]
@piranna 
> I still think this could be done as a npm module and don't have a direct relationship with runtime.js beyond the fact that it could provide some modules to access the direct hardware (that could also be npm modules available for other Javascript environments...).

Well, if this converting from software modules to hardware modules made correctly (in right way in right place, see -  [Intel unveils new Xeon chip with integrated FPGA, touts 20x performance boost](http://www.extremetech.com/extreme/184828-intel-unveils-new-xeon-chip-with-integrated-fpga-touts-20x-performance-boost)) this is the same, what V8 make with interpreted JS code, when compiled it, but done twice with twice effort :) Yes, I know, you may say that I'm too greedy, but why not? :D Anyway the common direction of hardware envelopment goes in to programmable hardware and universal periferal devices. So this is related to more common issue about what hardware is better first choice for runtime.js porting :) 

Another question - how is it difficult to make converting tools a part of runtime.js/nodeos. Yes, good question, I think it's not very easy, but I think it's possible :)

**Comment by danxn** [comment]
Anyway high-level programming languages now is very popular design tool for logical circuits:

http://esd.cs.ucr.edu/labs/tutorial/comparator.vhd

```code
---------------------------------------------------
-- n-bit Comparator (ESD book figure 2.5)
-- by Weijun Zhang, 04/2001 
--
-- this simple comparator has two n-bit inputs & 
-- three 1-bit outputs
---------------------------------------------------
 
library ieee;
use ieee.std_logic_1164.all;

---------------------------------------------------

entity Comparator is

generic(n: natural :=2);
port(	A:	in std_logic_vector(n-1 downto 0);
	B:	in std_logic_vector(n-1 downto 0);
	less:		out std_logic;
	equal:		out std_logic;
	greater:	out std_logic
);
end Comparator;

---------------------------------------------------

architecture behv of Comparator is

begin 
    
    process(A,B)
    begin
        if (A<B) then 
	    less <= '1';
	    equal <= '0';
	    greater <= '0';
	elsif (A=B) then   
	    less <= '0';
	    equal <= '1';
	    greater <= '0';
	else 
	    less <= '0';
	    equal <= '0';
	    greater <= '1';
	end if;
    end process;

end behv;

---------------------------------------------------

**Comment by danxn** [comment]
Even if you don't understand 中文 (like me) it's not a big problem, because video clearly explains Sandbox tools set for logic circuits prototyping [視窗程式：程式領域漫談 (Node.js, Verilog, Spice)](https://www.youtube.com/watch?v=GxW13uw856s)

**Comment by danxn** [comment]
Actually it's good idea - just create needed hardware platform and port need software to it and then test all new features. Anyway most computer devices interacting with help of networking protocols, so it is no matter what exactly software and hardware it has. Ok, need to make some investigations what is need to be done to create custom hardware PSoC card with newer version of Zynq.

**Comment by danxn** [comment]
Ok, actually this is also interesting, because you can avoid "drivers hell" for significant part of external devices, by implementing this devices programmatically, relying only on their standards and specifications, not on their models, produced by different vendors.

**Comment by iefserge** [comment]
@danxn Yeah, "drivers hell" is a an issue for any new OS, that's why it's easier to target hypervisor like KVM or Xen and just support the virtualized interfaces. You'd usually want to run your services in a cloud anyway instead of setting up a dedicated physical hardware. Specific hardware drivers could be written later if really necessary.
I'd love to see a standardized hardware interface supported by all vendors though.

**Comment by danxn** [comment]
@iefserge 
> I'd love to see a standardized hardware interface supported by all vendors though.

Yes, this is a big headache and standartized interfaces makes life of developers much easier.

For me in this direction more interesting using of FPGA modules as hardware accelerator of JS modules. Now I'm learning this and looks like custom logic hardware cores converted from JS modules is definetely reachable thing. Just interesting how they can speed-up server applications. Very interesting :) Now I'm looking for easy entrance point for this experiments. May be it will be not a runtime.js first, but Linux and C. But with JS OS and JS apps it will be much cooler :)
